packages/nouns-contracts/contracts/Rewards.sol

```diff
@@ -0,0 +1,280 @@
+        uint16 auctionRewardBps;
+        uint16 proposalEligibilityQuorumBps;
+    }
+
+    constructor(
+        address owner,
+        address nounsDAO_,
+        address auctionHouse_,
+        address ethToken_,
+        uint32 nextProposalIdToReward_,
+        uint256 lastProcessedAuctionId_,
+        uint256 nextProposalRewardFirstAuctionId_,
+        RewardParams memory rewardParams
+    ) Ownable(owner) {
+        nounsDAO = INounsDAOLogicV3(nounsDAO_);
+        auctionHouse = INounsAuctionHouseV2(auctionHouse_);
+        ethToken = IERC20(ethToken_);
+        nextProposalIdToReward = nextProposalIdToReward_;
+        lastProcessedAuctionId = lastProcessedAuctionId_;
+        nextProposalRewardFirstAuctionId = nextProposalRewardFirstAuctionId_;
+        params = rewardParams;
+    }
+
+    struct Temp {
+        uint256 numEligibleVotes;
+        uint256 rewardPerProposal;
+        uint256 rewardPerVote;
+    }
+
+    function setParams(RewardParams calldata newParams) public onlyOwner {
+        params = newParams;
+    }
+
+    function getParams() public view returns (RewardParams memory) {
+        return params;
+    }
+
+    function updateRewardsForAuctions(uint256 lastNounId) public {
+        uint256 currentlyAuctionedNounId = auctionHouse.auction().nounId;
+        require(lastNounId < currentlyAuctionedNounId, 'lastNounId must be settled');
+
+        uint256 lastProcessedAuctionId_ = lastProcessedAuctionId;
+        require(lastNounId > lastProcessedAuctionId_, 'lastNounId must be higher');
+        lastProcessedAuctionId = lastNounId;
+
+        INounsAuctionHouseV2.Settlement memory settlement;
+        for (uint256 nounId = lastProcessedAuctionId_ + 1; nounId <= lastNounId; nounId++) {
+            // TODO maybe can be optimized to one call with large range
+            settlement = auctionHouse.getSettlements(nounId, nounId + 1, false)[0];
+            if (settlement.clientId > 0) {
+                clientBalances[settlement.clientId] += (settlement.amount * params.auctionRewardBps) / 10_000;
+            }
+        }
+    }
+
+    /**
+     * @param lastProposalId id of the last proposal to include in the rewards distribution. all proposals up to and
+     * including this id must have ended voting.
+     * @param lastAuctionedNounId the noun id that was auctioned when proposal with id `lastProposalId` was created.
+     * The auction revenue up to this noun id are included in the bounty.
+     * @param votingClientIds array of client ids that were used to vote on of all eligible the eligible proposals in
+     * this rewards distribution
+     */
+    function updateRewardsForProposalWritingAndVoting(
+        uint32 lastProposalId,
+        uint256 lastAuctionedNounId,
+        uint32[] calldata votingClientIds
+    ) public {
+        require(lastProposalId <= nounsDAO.proposalCount(), 'bad lastProposalId');
+        require(lastProposalId >= nextProposalIdToReward, 'bad lastProposalId');
```

Comment:
worth caching `nextProposalIdToReward` into memory, I see it's currently read twice